cloudinit/sources/DataSourceVMware.py

```diff
@@ -685,20 +685,10 @@ def is_valid_ip_addr(val):
     Returns false if the address is loopback, link local or unspecified;
     otherwise true is returned.
     """
-    # TODO(extend cloudinit.net.is_ip_addr exclude link_local/loopback etc)
-    # TODO(migrate to use cloudinit.net.is_ip_addr)#
-
-    addr = None
-    try:
-        addr = ipaddress.ip_address(val)
-    except ipaddress.AddressValueError:
-        addr = ipaddress.ip_address(str(val))
-    except Exception:
-        return None
-
-    if addr.is_link_local or addr.is_loopback or addr.is_unspecified:
-        return False
-    return True
+    addr = net.maybe_get_address(ipaddress.ip_address, val)
```

Comment:
I don't think this is ultimately what we want here as this introduces a small cost to all `net.is_ip*` functions which now have to call the `maybe_get_address` function.

Ideally, I'd like to see `is_ip_addr` grow a `allow_unroutable: bool = True` which would only invoke the is_link_local, is_loopback, is_unspecified class methods if `True`. Then we can drop this function altogether and just call `net.is_ip_addr(ip_info["addr"], allow_unroutable=False)` down in VMWare's `get_host_info`.

Reply:
Do I understand the concern correctly that an additional function call overhead would be too great and that new kwargs+conditional in `is_ip_addr` won't add as much overhead to each call?

Function wrapper vs additional kwargs is not something I've measured, but intuition says this tradeoff smells of performance over-optimization. Maybe I'm wrong?

Reply:
The concern was that we were affecting all `is_ip_*` functions in order to tease out a need for the VMWare datasource that seemed like if could have been confined to just the `is_ip_address` function without any impact to the other utility functions. That said it was a minor concern and shouldn't block progress on this branch